SAAS-67 Document go logging facility and prepare it for release

kostrahb opened the pull request.

Shimmi commented on handlers/init/Initialization.go:
```diff
@@ -0,0 +1,68 @@
+
+	log.SetHandler(multi.New(handlers...))
+	log.SetLevelFromString(config.LogLevel)
+	logger = log.WithFields(config.Context)
```
forgot to call `setAdditionalData`?

kostrahb committed:
[SAAS-66] Documentation and code formatting

kostrahb committed:
[SAAS-67] Documented code, removed some handlers

Removed handlers which have fixed number of levels and don't support custom ones.

kostrahb committed:
[SAAS-67] Readme, licence and some code style

kostrahb committed:
[SAAS-67] Fixed licence, updated readme

Shimmi commented on _examples/json/json.go:
```diff
@@ -31,6 +31,6 @@ func main() {
 		ctx.Warn("upload retry")
 		one(ctx)
-//		ctx.WithError(errors.New("unauthorized")).Error("upload failed")
+		//		ctx.WithError(errors.New("unauthorized")).Error("upload failed")
```
single mezera only
```suggestion
		// ctx.WithError(errors.New("unauthorized")).Error("upload failed")
```

Shimmi commented on Readme.md:
```diff
@@ -1,29 +1,135 @@
+# Go logging facility
 
-![Structured logging for golang](assets/title.png)
+Based on [Apex log](https://github.com/apex/log) with few tweaks. Unless otherwise noted (mainly in `pkg.go`), this package should be thread safe. Just for the record the architecture leaves almost all thread safety on handlers so if you are uncertain, consult the documentation and code of handlers you use.
```
```suggestion
Based on [Apex log](https://github.com/apex/log) with few tweaks. Unless otherwise noted (mainly in `pkg.go`), this package should be thread safe. Just for the record the architecture leaves almost all thread safety on handlers so if you are uncertain, consult the documentation and code of the handlers you use.
```

Shimmi commented on Readme.md:
```diff
@@ -1,29 +1,135 @@
+
+### Autogenerated levels
+
+This package contains python script for log level automatic generation. The script contains default values used in SOCIFI but can be easily tweaked to generate entirely custom level names and level numbers. Just alter the level names and level codes at the top of the file `gen_levels.py` and run it. It will take care about everything else (unless you deleted guideline comments).
```
Would not use "The script contains default values used in SOCIFI" but rather "[RFC 5424](https://tools.ietf.org/html/rfc5424)". It's more generic and looks more professional :)
```suggestion
This package contains python script for log level automatic generation. By default, the script uses log levels that comply with RFC 5424 levels (debug, info, notice, warning, error, critical, alert, emergency) but can be easily tweaked to generate entirely custom level names and level numbers. Just alter the level names and level codes at the top of the file `gen_levels.py` and run it. It will take care about everything else (unless you deleted guideline comments).
```

Shimmi commented on Readme.md:
```diff
@@ -1,29 +1,135 @@
+
+### Hooks
+
+This package allows registering of hooks which will be run for every field in entry to check specific things in field name and then apply some function to it's content. It's incredibly useful e.g. for password hiding in logs. Example implementations of hooks are in folder `hook/` and amongst them is a dsn hook which will check if a field name contains `dsn` (case insensitive). This package will t ...
```
```suggestion
This package allows registering of hooks which will be run for every field in an entry to check specific things in field name and then apply some function to its content. It's incredibly useful e.g. for password hiding in logs. Example implementations of hooks are in folder `hook/` and amongst them is a dsn hook which will check if a field name contains `dsn` (case insensitive). This package will then try to convert dsn interface to string and hide password in it. The rationale behind this is that some information might be very useful in troubleshooting some problems but logging passwords is very bad practice. This feature is entirely opt-in and might be included by importing module dsn as such:
```

Shimmi replied:
I would split it to make the message more clear:
1) It supports hooks (what is it, how to use it, ...)
2) DNS string sanitizer (what it is, why should one use it, how, ...)

Generally, try to avoid such long text. The text shall be as short and as easy to read as possible. Long texts are more likely to be skipped when reading. (at least use paragraphs :-) )

Shimmi commented on Readme.md:
```diff
@@ -1,29 +1,135 @@
+
+### Atexit
+
+Inspired by [atexit module](github.com/tebeka/atexit) go logging facility also offers possibility to register a function which will be run on system exit. However this functionality has the same caveat as atexit, the program needs to closed with any `Exit(code int)` function in this module. This function as well as `AddExitHandler(handler func())` which registers functions to be run on program ex ...
```
```suggestion
Inspired by [atexit module](github.com/tebeka/atexit) go logging facility also offers the possibility to register a function which will be run on system exit. However this functionality has the same caveat as atexit, the program needs to be closed with any `Exit(code int)` function in this module. This function as well as `AddExitHandler(handler func())` which registers functions to be run on program exit is defined for both module, logger and entry.
```

Shimmi commented on Readme.md:
```diff
@@ -1,29 +1,135 @@
+
+### Simple logger initialization from config json
+
+There is also initialization script which allows simple load of configuration needed by logger. This feature is now limited to json and elastic handlers and might be changed in the future to allow better interoperability with e.g. [Viper](https://github.com/spf13/viper). __We therefore cannot guarantee any backward compatibility in this module!__
```
```suggestion
There is also initialization script which allows a simple load of configuration needed by the ogger. This feature is now limited to JSON and elastic handlers and might be changed in the future to allow better interoperability with e.g. [Viper](https://github.com/spf13/viper). __We, therefore, cannot guarantee any backward compatibility in this module!__
```

Shimmi commented on Readme.md:
```diff
@@ -1,29 +1,135 @@
+### Simple logger initialization from config json
+
+There is also initialization script which allows simple load of configuration needed by logger. This feature is now limited to json and elastic handlers and might be changed in the future to allow better interoperability with e.g. [Viper](https://github.com/spf13/viper). __We therefore cannot guarantee any backward compatibility in this module!__
+But for now the example usage is following:
```
```suggestion
But for now the example usage is the following:
```

Shimmi commented on Readme.md:
```diff
@@ -1,29 +1,135 @@
 ## Handlers
 
-- __cli__ – human-friendly CLI output
+Some handlers which support only fixed number of log levels have been discarded and only these are left. PR for simple text handler which is able to process any number of levels is welcome.
```
```suggestion
Some handlers which support a fixed number of log levels only have been discarded and only the following were kept. PR for simple text handler which will be able to process any number of levels is welcome.
```

kostrahb committed:
[SAAS-67] Improved documentation
